File: src/gui/Window.java
package gui;

import javafx.application.Application;
import javafx.scene.Scene;
import javafx.scene.control.Label;
import javafx.scene.layout.*;
import javafx.scene.paint.Paint;
import javafx.stage.Stage;

public class Window extends Application
{
    private static final String WINDOW_TITLE = "Window Title";
    private static final int WINDOW_WIDTH = 640;
    private static final int WINDOW_HEIGHT = 480;

    public static void main(String[] args)
    {
        launch(args);
    }

    @Override
    // stage: the window
    public void start(Stage primaryStage) throws Exception
    {
        // configure the stage stage (the window)
        primaryStage.setTitle(WINDOW_TITLE);
        // set the scene (inside the window)
        primaryStage.setScene(new Scene(new ContentPane(),WINDOW_WIDTH,WINDOW_HEIGHT));
        primaryStage.show();
    }

    private class ContentPane extends BorderPane
    {
        public ContentPane()
        {
            super();

            // add a red, dashed 3 pixel, rounded thick border about the layout
            setBorder(new Border(new BorderStroke(Paint.valueOf("f00"),BorderStrokeStyle.DASHED,new CornerRadii(10),new BorderWidths(3))));

            // ForwardingPane
            setCenter(new ForwardingPane());

            // SettingsPane, StatisticsPane
            GridPane bottom = new GridPane();
            bottom.setGridLinesVisible(true);

            ColumnConstraints constraints = new ColumnConstraints();
            constraints.setFillWidth(true);
            constraints.setHgrow(Priority.ALWAYS);
            bottom.getColumnConstraints().add(constraints);
            bottom.getColumnConstraints().add(constraints);

            bottom.add(new SettingsPane(),0,0);
            bottom.add(new StatisticsPane(),1,0);
            setBottom(bottom);
        }
    }
}

File: peer/src/gui/Window.kt
package gui

import ProtocolPeer.ClientSocket
import ProtocolPeer.Connection
import javafx.application.Application
import javafx.beans.InvalidationListener
import javafx.scene.Scene
import javafx.scene.control.ScrollPane
import javafx.scene.layout.BorderPane
import javafx.stage.Stage

class Window:Application()
{
    private val WINDOW_TITLE = "Window Title"

    private val WINDOW_WIDTH = 1000
    private val WINDOW_HEIGHT = 480

    private val connectionsPane = ConnectionsPane()
    private val serverPane = ServerPane()
    private val clientSocketObserver = ClientSocketObserver()

    private var clientSocket:ClientSocket? = null

    override fun start(primaryStage:Stage)
    {
        // configure the stage stage (the window)
        primaryStage.title = WINDOW_TITLE

        // set the scene (inside the window)
        primaryStage.scene = Scene(ContentPane(),WINDOW_WIDTH.toDouble(),WINDOW_HEIGHT.toDouble())
        primaryStage.scene.stylesheets.add(CSS.FILE_PATH)
        primaryStage.show()

        serverPane.port.addListener(InvalidationListener
        {
            clientSocket = ClientSocket(serverPane.port.value,clientSocketObserver)
            clientSocket!!.setListening(serverPane.listening.value)
        })
        serverPane.listening.addListener(InvalidationListener
        {
            clientSocket?.setListening(serverPane.listening.value)
        })
    }

    private inner class ClientSocketObserver:ClientSocket.Observer
    {
        override fun onAccept(newConnection:Connection?)
        {
            // todo: implement
            throw UnsupportedOperationException()
        }
    }

    private inner class ContentPane:BorderPane()
    {
        init
        {
            // populate the content pane
            center = ScrollPane(connectionsPane)
            bottom = serverPane
        }
    }
}
